[SPARK-13139][SQL] Parse Hive DDL commands ourselves

andrewor14 opened the pull request:
## What changes were proposed in this pull request?

This patch is ported over from @viirya's changes in #11048. Currently for most DDLs we just pass the query text directly to Hive. Instead, we should parse these commands ourselves and in the future (not part of this patch) use the `HiveCatalog` to process these DDLs. This is a pretext to merging `SQLContext` and `HiveContext`.

Note: As of this patch we still pass the query text to Hive. The difference is that we now parse the commands ourselves so in the future we can just use our own catalog.
## How was this patch tested?

Jenkins, new `DDLCommandSuite`, which comprises of about 40% of the changes here.

Commit:
Move things into new ParserUtils object

Commit:
Merge branch 'master' of github.com:apache/spark into parser-plus-plus

Commit:
Port over viirya's changes in #11048

Commit:
Address comments from #11408 + fix style

Commit:
Merge branch 'master' of github.com:apache/spark into parser-plus-plus

Conflicts:
	sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/bucket.scala

Commit:
Minor fixes

andrewor14 commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala:
```diff
@@ -0,0 +1,197 @@
+    buckets: Option[BucketSpec],
+    // TODO: use `clustered` and `sorted` instead for simplicity
+    noClustered: Boolean,
+    noSorted: Boolean)(sql: String)
```
@viirya was there any reason why these have to be negative? It's much easier to understand if it's positive, i.e. `clustered` and `sorted`.

viirya replied:
Just because the corresponding token is `TOK_NOT_CLUSTERED` and `TOK_NOT_SORTED`. We can use positive here.

SparkQA commented:
**[Test build #52630 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52630/consoleFull)** for PR 11573 at commit [`a663b5c`](https://github.com/apache/spark/commit/a663b5ccca7718b41df8b8ede462ac6c9e0b8e8f).

andrewor14 commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala:
```diff
@@ -0,0 +1,197 @@
+    storedAsDirs: Boolean,
+    notSkewed: Boolean,
+    // TODO: what??
+    notStoredAsDirs: Boolean)(sql: String)
```
@viirya I'm really confused about this flag. We have another one called `storedAsDirs` and this one says `notStoredAsDirs`. It would seem that one is always the opposite of the other, but that's not actually the case in tests, e.g.

```
// DDLCommandSuite, "alter table: skewed"
val expected3 = AlterTableSkewed(
  TableIdentifier("table_name", None),
  Seq("dt", "country"),
  Seq(List("2008-08-08", "us"), List("2009-09-09", "uk")),
  storedAsDirs = false,
  notSkewed = false,
  notStoredAsDirs = false)(sql3)
```
seems contradictory?

viirya replied:
`notStoredAsDirs` is used when the command is `ALTER TABLE table_name NOT STORED AS DIRECTORIES`. When it is true, `storedAsDirs` is false of course.

`storedAsDirs` is used with skewed arguments columns and values.

I think we can use just one variable for them.

When the command is `ALTER TABLE table_name NOT SKEWED`, then the values of  `notStoredAsDirs` and `storedAsDirs` don't matter.

andrewor14 replied:
I've opted for having separate case classes for those commands, e.g. `AlterTableNotSkewed`. Trying to express them using `AlterTableSkewed` is pretty messy.

andrewor14 commented:
Note: The only changes I made on top of #11048 is addressing the outstanding comments in that patch and some minor clean ups. It's entirely possible that there still are things that are missing or incorrect given the original patch was not reviewed completely yet.

@hvanhovell @yhuai PTAL.

SparkQA commented:
**[Test build #52630 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52630/consoleFull)** for PR 11573 at commit [`a663b5c`](https://github.com/apache/spark/commit/a663b5ccca7718b41df8b8ede462ac6c9e0b8e8f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52630/
Test PASSed.

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkQl.scala:
```diff
@@ -64,6 +85,53 @@ private[sql] class SparkQl(conf: ParserConf = SimpleParserConf()) extends Cataly
 
+      case Token("TOK_CREATEDATABASE", Token(databaseName, Nil) :: createDatabaseArgs) =>
+        val Seq(
+          allowExisting,
```
Isn't `allowExists` the exact opposite of `IF NOT EXISTS`? I am asking because I have a similar case in my PR.

viirya replied:
Not quite understand your question?

hvanhovell replied:
When i see a parameter that is named `allowExists` I think the existence of the current table is allowed, and that as a consequence the current create table command is going to overwrite the existing table. Why not name this `ifNotExists`?

andrewor14 replied:
I've renamed it.

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkQl.scala:
```diff
@@ -64,6 +85,53 @@ private[sql] class SparkQl(conf: ParserConf = SimpleParserConf()) extends Cataly
+        }.toMap
+        CreateDatabase(databaseName, allowExisting.isDefined, location, comment, props)(node.source)
+
+      case Token("TOK_CREATEFUNCTION", func :: as :: createFuncArgs) =>
```
I am not sure this pattern works for functions with a dotted name. For example

    create function s.t as 'foo.bar.func'

Produces the following AST:

    TOK_CREATEFUNCTION 1, 0, 10, 16
    :- s 1, 4, 4, 16
    :- t 1, 6, 6, 18
    +- 'foo.bar.func' 1, 10, 10, 23

Notice how the name is splitted into two top-level ASTNodes.

I think we need to change the parser rule here.

andrewor14 replied:
will fix later

andrewor14 replied:
fixed in latest commit.

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AlterTableCommandParser.scala:
```diff
@@ -0,0 +1,410 @@
+  }
+
+  private def cleanAndUnquoteString(s: String): String = {
+    cleanIdentifier(unquoteString(s))
```
I think unqoute string does the cleaning (i.e. stripping leading and trailing quotes) for us.

andrewor14 replied:
if you look at `ParserUtils` that's done in `cleanIdentifier`

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AlterTableCommandParser.scala:
```diff
@@ -0,0 +1,410 @@
+        AlterTableMerge(tableIdent, partition)(node.source)
+
+      case Token("TOK_ALTERTABLE_RENAMECOL", args) :: _ =>
+        val oldName = args(0).text
```
Pattern match this?

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AlterTableCommandParser.scala:
```diff
@@ -0,0 +1,410 @@
+
+      case Token("TOK_ALTERTABLE_SKEWED_LOCATION",
+      Token("TOK_SKEWED_LOCATIONS",
+      Token("TOK_SKEWED_LOCATION_LIST", locationMaps) :: Nil) :: Nil) :: _ =>
```
Isn't the wildcard used in the previous match preventing this from ever happening? Why not swap them?

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AlterTableCommandParser.scala:
```diff
@@ -0,0 +1,410 @@
+          notStoredAsDirs = true)(node.source)
+
+      case (tableSkewed @ Token("TOK_ALTERTABLE_SKEWED", _)) :: _ =>
+        val skewedArgs = tableSkewed match {
```
This pattern match is redundant.

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AlterTableCommandParser.scala:
```diff
@@ -0,0 +1,410 @@
+          notStoredAsDirs = false)(node.source)
+
+      case Token("TOK_ALTERTABLE_SKEWED_LOCATION",
+      Token("TOK_SKEWED_LOCATIONS",
```
Style

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AlterTableCommandParser.scala:
```diff
@@ -0,0 +1,410 @@
+      tableIdent: TableIdentifier,
+      partition: Option[Map[String, Option[String]]]): LogicalPlan = {
+    nodes match {
+      case rename @ Token("TOK_ALTERTABLE_RENAME", renameArgs) :: _ =>
```
Whe the wildcards at the end every pattern match. Do we expect additional input which can safely ignore?

andrewor14 replied:
@viirya can you answer this?

viirya replied:
Ah. That is because the node is structured as (take `TOK_ALTERTABLE_RENAME` as an example):

    Token("TOK_TABNAME", _) :: Token("TOK_ALTERTABLE_RENAME", renameArgs) :: Token("TOK_PARTSPEC", _)

As you see, Token("TOK_PARTSPEC", _) is put at the end of the AST. Although I've parsed partspec early, I pass the all nodes after `Token("TOK_TABNAME", _)` to the method `matchAlterTableCommands`. So there is always an additional input that can safely ignore.

viirya replied:
If you extract the command from the AST, we can remove the additional input of course.

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AlterTableCommandParser.scala:
```diff
@@ -0,0 +1,410 @@
+        val partitions: ArrayBuffer[(Map[String, Option[String]], Option[String])] =
+          new ArrayBuffer()
+        var currentPart: Map[String, Option[String]] = null
+        parts.map {
```
We could change the parser rule in such a way that it will nest the locations inside a partition specification.

andrewor14 replied:
we could do that later

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AlterTableCommandParser.scala:
```diff
@@ -0,0 +1,410 @@
+        val Some(newPartition) = parsePartitionSpec(partArg)
+        AlterTableRenamePartition(tableIdent, partition.get, newPartition)(node.source)
+
+      case Token("TOK_ALTERTABLE_EXCHANGEPARTITION",
```
We assume that `TOK_ALTERTABLE_EXCHANGEPARTITION` is followed by `TOK_PARTSPEC` and `TOK_TABNAME`. Why not let `parsePartitionSpec` and `extractTableIdent` validate this for us? This would save us some trouble in the match.

andrewor14 replied:
agreed

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AlterTableCommandParser.scala:
```diff
@@ -0,0 +1,410 @@
+
+      case _ =>
+        throw new AnalysisException(
+          s"Unexpected children nodes in alter table command: '${node.text}'")
```
Error message could be better, e.g.: `Unknown ALTER TABLE command: ....`?

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/bucket.scala:
```diff
@@ -29,7 +32,8 @@ package org.apache.spark.sql.execution.datasources
     bucketColumnNames: Seq[String],
-    sortColumnNames: Seq[String])
+    sortColumnNames: Seq[String],
+    sortDirections: Seq[SortDirection] = Nil)
```
I still think sortColumnNames and sortDirections should be combined into a seq of pairs.

andrewor14 replied:
will fix later

andrewor14 replied:
fixed in latest commit.

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala:
```diff
@@ -0,0 +1,197 @@
+
+
+/**
+ * A DDL command expected to be run in the underlying system without Spark parsing the
```
The query has been parsed.... it should be something like ...without Spark executing the querty...

Commit:
Merge branch 'master' of github.com:apache/spark into parser-plus-plus

Commit:
Rename `notClustered` and `notSorted`

Commit:
Significant cleanup + documentation in ALTER TABLE

This commit cleans up the case class signatures representing
ALTER TABLE commands along with the code that parses them.
This commit also adds a lot of missing documentation that
was sorely needed for AlterTableCommandParser to be readable.

The cleanup in this commit concerns only ALTER TABLE commands,
but is expected to spread to other parts of the PR in subsequent
commits.

SparkQA commented:
**[Test build #52737 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52737/consoleFull)** for PR 11573 at commit [`f54501f`](https://github.com/apache/spark/commit/f54501f6c244c8e37b332991fd22617ec1ef1e76).

SparkQA commented:
**[Test build #52739 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52739/consoleFull)** for PR 11573 at commit [`6538829`](https://github.com/apache/spark/commit/6538829fea21463826970997cf05eaedfcadb100).

andrewor14 commented:
@hvanhovell I've addressed most of your comments except a couple ones where I said I would fix later. Separately I've also significantly cleaned up the logic and signatures in the alter table code so you should have a look at that as well. It's probably quite different from what it looked like when you last reviewed this patch.

SparkQA commented:
**[Test build #52741 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52741/consoleFull)** for PR 11573 at commit [`6c3f994`](https://github.com/apache/spark/commit/6c3f9946c5244bc61776183f412acb7be068795a).

SparkQA commented:
**[Test build #52737 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52737/consoleFull)** for PR 11573 at commit [`f54501f`](https://github.com/apache/spark/commit/f54501f6c244c8e37b332991fd22617ec1ef1e76).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `abstract class NativeDDLCommand(val sql: String) extends RunnableCommand `
  * `case class AlterTableUnsetProperties(`
  * `case class AlterTableStorageProperties(`
  * `case class AlterTableNotClustered(`
  * `case class AlterTableNotSorted(`
  * `case class AlterTableNotSkewed(`
  * `case class AlterTableNotStoredAsDirs(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52737/
Test FAILed.

SparkQA commented:
**[Test build #52739 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52739/consoleFull)** for PR 11573 at commit [`6538829`](https://github.com/apache/spark/commit/6538829fea21463826970997cf05eaedfcadb100).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `abstract class NativeDDLCommand(val sql: String) extends RunnableCommand `
  * `case class AlterTableUnsetProperties(`
  * `case class AlterTableStorageProperties(`
  * `case class AlterTableNotClustered(`
  * `case class AlterTableNotSorted(`
  * `case class AlterTableNotSkewed(`
  * `case class AlterTableNotStoredAsDirs(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52739/
Test FAILed.

SparkQA commented:
**[Test build #52741 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52741/consoleFull)** for PR 11573 at commit [`6c3f994`](https://github.com/apache/spark/commit/6c3f9946c5244bc61776183f412acb7be068795a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `abstract class NativeDDLCommand(val sql: String) extends RunnableCommand `
  * `case class AlterTableUnsetProperties(`
  * `case class AlterTableStorageProperties(`
  * `case class AlterTableNotClustered(`
  * `case class AlterTableNotSorted(`
  * `case class AlterTableNotSkewed(`
  * `case class AlterTableNotStoredAsDirs(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52741/
Test FAILed.

Commit:
Fix tests

SparkQA commented:
**[Test build #52760 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52760/consoleFull)** for PR 11573 at commit [`794ae78`](https://github.com/apache/spark/commit/794ae7854291abb3b7558b1465c9385fb29cc22c).

SparkQA commented:
**[Test build #52760 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52760/consoleFull)** for PR 11573 at commit [`794ae78`](https://github.com/apache/spark/commit/794ae7854291abb3b7558b1465c9385fb29cc22c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52760/
Test PASSed.

Commit:
Fix CREATE FUNCTION with dot + clean up SparkQl

This used to fail: CREATE FUNCTION dbname.fname AS alias.
Now it doesn't!

Commit:
Group (col_name, sort_dir) in BucketSpec

Commit:
Merge branch 'master' of github.com:apache/spark into parser-plus-plus

andrewor14 commented:
@hvanhovell I believe as of the latest commit I've addressed all of your comments. This is ready from my side so I've removed the WIP tag. PTAL.

SparkQA commented:
**[Test build #52770 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52770/consoleFull)** for PR 11573 at commit [`37854fc`](https://github.com/apache/spark/commit/37854fccec3af29f15acfdd4693a532b6f862469).

Commit:
Minor changes

SparkQA commented:
**[Test build #52773 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52773/consoleFull)** for PR 11573 at commit [`6ad8dd5`](https://github.com/apache/spark/commit/6ad8dd5f5294fb7c9424651920706645d1dba8f9).

SparkQA commented:
**[Test build #52773 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52773/consoleFull)** for PR 11573 at commit [`6ad8dd5`](https://github.com/apache/spark/commit/6ad8dd5f5294fb7c9424651920706645d1dba8f9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52773/
Test PASSed.

andrewor14 commented:
Also cc @rxin @yhuai.

SparkQA commented:
**[Test build #52770 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52770/consoleFull)** for PR 11573 at commit [`37854fc`](https://github.com/apache/spark/commit/37854fccec3af29f15acfdd4693a532b6f862469).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52770/
Test PASSed.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/bucket.scala:
```diff
@@ -24,12 +27,15 @@ package org.apache.spark.sql.execution.datasources
     numBuckets: Int,
     bucketColumnNames: Seq[String],
-    sortColumnNames: Seq[String])
+    sortColumns: Seq[(String, SortDirection)]) {
```
It is great to have `SortDirection` defined explicitly. So, when we create it we know what will the order by.

However, this change implies that we can use `descending` as the order, which is actually not allowed because `UnsafeKVExternalSorter` only sort keys with ascending order (this sorter is used in `DynamicPartitionWriterContainer` when we generate bucketing files). So, I think for now, it is better to revert this change. In future, we can revisit it when we store sort direction in metastore and we can actually sort rows in a bucket file with a descending order.

Commit:
Merge branch 'master' of github.com:apache/spark into parser-plus-plus

Commit:
Revert "Group (col_name, sort_dir) in BucketSpec"

This reverts commit 307a588009dcd8479fe219157b25a8c4b420c126.

Commit:
Revert sort directions in bucket spec for now

SparkQA commented:
**[Test build #52937 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52937/consoleFull)** for PR 11573 at commit [`bd91b0f`](https://github.com/apache/spark/commit/bd91b0f3c850da796b22c7e575658625d2486057).

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkQl.scala:
```diff
@@ -64,10 +83,86 @@ private[sql] class SparkQl(conf: ParserConf = SimpleParserConf()) extends Cataly
+            }
+          case _ => parseFailed("Invalid CREATE FUNCTION command", node)
+        }.toMap
+        CreateFunction(funcName, alias, resourcesMap, temp.isDefined)(node.source)
```
What will this map look like if I have `USING JAR 'jar1', JAR 'jar2', ...`?

yhuai replied:
Let's also have a test for this case.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkQl.scala:
```diff
@@ -64,10 +83,86 @@ private[sql] class SparkQl(conf: ParserConf = SimpleParserConf()) extends Cataly
+        }
+        val props = dbprops.toSeq.flatMap {
+          case Token("TOK_DATABASEPROPERTIES", Token("TOK_DBPROPLIST", propList) :: Nil) =>
+            extractProps(propList, "TOK_TABLEPROPERTY")
```
Maybe it will be also good to provide an example format like what we have for `TOK_CREATEFUNCTION`?

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkQl.scala:
```diff
@@ -29,7 +29,26 @@ private[sql] class SparkQl(conf: ParserConf = SimpleParserConf()) extends Cataly
+  /**
+   * For each node, extract properties in the form of a list ['key1', 'key2', 'key3', 'value']
+   * into a pair (key1.key2.key3, value).
+   */
```
Maybe it is good to provide a concrete example as well?

yhuai replied:
Also, what is the difference between this method and `extractTableProps` defined for alter table?

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AlterTableCommandParser.scala:
```diff
@@ -0,0 +1,428 @@
+    node match {
+      case Token("TOK_PARTSPEC", partitions) =>
+        partitions.map {
+          // Note: sometimes there's a "=", "<" or ">" between the key and the value
```
Maybe it will be good to provide a concrete example? Is it caused by what a user type?

yhuai commented:
I have left a few comments. It is a good starting point. Thank you for working on this!

SparkQA commented:
**[Test build #52937 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52937/consoleFull)** for PR 11573 at commit [`bd91b0f`](https://github.com/apache/spark/commit/bd91b0f3c850da796b22c7e575658625d2486057).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52937/
Test PASSed.

yhuai commented:
OK. Let's merge this first (to avoid it has conflicts caused by other commits ). We can address the comments in a follow-up PR.
